### PR TITLE
Scale the playfield using the diagonal instead of the max side when barrel roll mod is active

### LINKED ISSUE #13683: Hit circles can appear partially off-screen in certain resolutions with barrel roll mod
**Describe the bug:** Seems to only be a problem for hit circles placed near a corner of the playfield.

**Screenshots or videos showing encountered issue:** 

https://user-images.githubusercontent.com/55509723/123685385-7c7cfe80-d803-11eb-8768-f5ced57b78ec.mp4

Beatmap in video: 7 star map of [Chasers - Lost (ktgster).zip](https://github.com/ppy/osu/files/6728472/Chasers.-.Lost.ktgster.zip)

**osu!lazer version:** commit a9084db6659840f1d074aec0e74168786f36e386

### PR DESCRIPTION
**Pull request by aitani9**
Closes #13683.

Example of behavior with changes:

https://user-images.githubusercontent.com/55509723/126569332-e4549ff7-d2ca-4e95-9170-4b3c8456e58b.mp4

### CODE REVIEW

File: osu.Game/Rulesets/Mods/ModBarrelRoll.cs [hunk 1]
```diff
@@ -50,8 +50,8 @@ public void ApplyToDrawableRuleset(DrawableRuleset<TObject> drawableRuleset)
 
             var playfieldSize = drawableRuleset.Playfield.DrawSize;
             var minSide = MathF.Min(playfieldSize.X, playfieldSize.Y);
-            var maxSide = MathF.Max(playfieldSize.X, playfieldSize.Y);
-            drawableRuleset.Playfield.Scale = new Vector2(minSide / maxSide);
+            var diagonal = MathF.Sqrt(MathF.Pow(playfieldSize.X, 2) + MathF.Pow(playfieldSize.Y, 2));
+            drawableRuleset.Playfield.Scale = new Vector2(minSide / diagonal);
```
Comment (frenzibyte) [inline osu.Game/Rulesets/Mods/ModBarrelRoll.cs:54]: This is what `playfieldSize.Length` would do, which even has a faster version (`playfieldSize.LengthFast`) that is already used in `OsuModRandom`.

**Review by frenzibyte — COMMENTED** [review]
Seems fine otherwise.

File: osu.Game/Rulesets/Mods/ModBarrelRoll.cs (hunk 1 above)
Comment (frenzibyte) [inline osu.Game/Rulesets/Mods/ModBarrelRoll.cs:54]:
```suggestion
            drawableRuleset.Playfield.Scale = new Vector2(minSide / playfieldSize.LengthFast);
```

**Review by frenzibyte — CHANGES_REQUESTED** [review]

File: osu.Game/Rulesets/Mods/ModBarrelRoll.cs (hunk 1 above)
Comment (bdach) [reply osu.Game/Rulesets/Mods/ModBarrelRoll.cs:54]: Not sure why go for `LengthFast`, it's not exactly a hot path - this gets calculated and set once, so may as well do normal `Length` - but whichever works I guess...?

Comment (frenzibyte) [reply osu.Game/Rulesets/Mods/ModBarrelRoll.cs]: It's randomly used across the game which confused me, either way works anyways.

**Review by frenzibyte — APPROVED** [review]

**Comment by peppy** [comment]
I'm still not sure about this. The playfield is so small the game will likely be unplayable on smaller devices. @smoogipoo thoughts?

**Comment by smoogipoo** [comment]
It's more correct but you do have a point.

**Comment by aitani9** [comment]
I'm not sure if there's a way to get around making the playfield smaller, since the issue is that some hit circles can appear off-screen. If playability is a concern, then I would question whether the mod should exist to begin with.

**Comment by frenzibyte** [comment]
Yeah I think this is an unavoidable thing and would have to be counted as part of the mod's difficulty.

**Comment by peppy** [comment]
I don't think the reduction needs to be as high as it is. I'll do some further experimentation because I'm not happy with this direction.

**Comment by smoogipoo** [comment]
Closing this as ~1 year out of date, with some objecting to the current implementation.

**Pull request closed without merging**